hal/src/trackerm/network/network.cpp

```diff
@@ -0,0 +1,260 @@
+    std::unique_ptr<WifiNcpClient> ncpClient_;
+    std::unique_ptr<WifiNetworkManager> mgr_;
+
+    int init() {
+        // Initialize NCP client
+        std::unique_ptr<WifiNcpClient> ncpClient(new(std::nothrow) RealtekNcpClient);
+        CHECK_TRUE(ncpClient, SYSTEM_ERROR_NO_MEMORY);
+        auto conf = NcpClientConfig()
+                .eventHandler(RealtekNcpNetif::ncpEventHandlerCb, wl4)
+                .dataHandler(RealtekNcpNetif::ncpDataHandlerCb, wl4);
+        CHECK(ncpClient->init(std::move(conf)));
+        // Initialize network manager
+        mgr_.reset(new(std::nothrow) WifiNetworkManager(ncpClient.get()));
+        CHECK_TRUE(mgr_, SYSTEM_ERROR_NO_MEMORY);
+        ncpClient_ = std::move(ncpClient);
+        return 0;
+    }
+};
+
+bool netifCanForwardIpv4(netif* iface) {
+    if (iface && netif_is_up(iface) && netif_is_link_up(iface)) {
+        auto addr = netif_ip_addr4(iface);
+        auto mask = netif_ip_netmask4(iface);
+        auto gw = netif_ip_gw4(iface);
+        if (!ip_addr_isany(addr) && !ip_addr_isany(mask) && !ip_addr_isany(gw)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+
+class CellularNetworkManagerInit {
+public:
+    CellularNetworkManagerInit() {
+        const int r = init();
+        SPARK_ASSERT(r == 0);
+    }
+
+    CellularNetworkManager* instance() const {
+        return mgr_.get();
+    }
+
+private:
+    std::unique_ptr<CellularNcpClient> client_;
+    std::unique_ptr<CellularNetworkManager> mgr_;
+
+    int init() {
+        // Get active SIM card
+        SimType sim = SimType::INVALID;
+        CHECK(CellularNetworkManager::getActiveSim(&sim));
+        CellularNcpClientConfig conf;
+        conf.simType(sim);
+        conf.ncpIdentifier(platform_primary_ncp_identifier());
+        conf.eventHandler(PppNcpNetif::ncpEventHandlerCb, pp3);
+        conf.dataHandler(PppNcpNetif::ncpDataHandlerCb, pp3);
+        // Initialize NCP client
+        std::unique_ptr<CellularNcpClient> client;
+        client.reset(new(std::nothrow) QuectelNcpClient);
+        CHECK_TRUE(client, SYSTEM_ERROR_NO_MEMORY);
+        CHECK(client->init(conf));
+        // Initialize network manager
+        mgr_.reset(new(std::nothrow) CellularNetworkManager(client.get()));
+        CHECK_TRUE(mgr_, SYSTEM_ERROR_NO_MEMORY);
+        client_ = std::move(client);
+        return 0;
+    }
+};
+
+} // unnamed
+
+WifiNetworkManager* wifiNetworkManager() {
+    static WifiNetworkManagerInit mgr;
+    return mgr.instance();
+}
+
+CellularNetworkManager* cellularNetworkManager() {
+    static CellularNetworkManagerInit mgr;
+    return mgr.instance();
+}
+
+} // particle
+
+int if_init_platform(void*) {
+    /* lo0 (created by LwIP) */
+
+    /* th1 - OpenThread (Deprecated) */
+    reserve_netif_index();
+
+    /* en2 - Ethernet FeatherWing (optional) */
+    uint8_t mac[6] = {0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff};
+    {
+        // const uint32_t lsb = __builtin_bswap32(NRF_FICR->DEVICEADDR[0]);
```

Comment:
This needs to be initialized with the base mac address that is populated into efuse on RTL872X-based platforms. Same needs to be applied to P2 platform. The easiest way to deal with this on P2 and Tracker M is to `hal_get_device_id`, the last 6 bytes are the mac address. For Ethernet interface we should increase the last byte by 3. No other modifications to mac address need to be applied (e.g. dropping multicast or locally-administered bit).